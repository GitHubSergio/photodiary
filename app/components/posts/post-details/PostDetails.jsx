// Imports
import React, { useEffect, useState } from 'react';
import { Text, View, ScrollView, Image } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import {
  useRoute,
  useNavigation,
  useIsFocused,
} from '@react-navigation/native';
import postDetailsStyles from './postDetailsStyles';
import moment from 'moment';

// Actions
import {
  getPostDetails,
  clearPostsDetails,
  isEditingPost,
} from '../../../redux/actions/postsActions';
import { deletePost } from '../../../redux/middlewares/postsMiddleWare';
import { alertFactory } from '../../shared/alert/alertFactory';

// Components
import Loader from '../../shared/loader/Loader';
import PostDetailsMenu from '../post-details-menu/PostDetailsMenu';

const PostDetails = () => {
  const { postDetails } = useSelector((state) => state.posts);
  const { allPosts } = useSelector((state) => state.posts);
  const { deletePostSuccess } = useSelector((state) => state.posts);

  const [progressiveImage, setProgressiveImageState] = useState(false);
  const dispatch = useDispatch();
  const route = useRoute();
  const navigation = useNavigation();
  const isFocused = useIsFocused();
  const currentDate = moment(postDetails.createdAt).format('dddd Do MMM YYYY');

  useEffect(() => {
    if (isFocused) {
      dispatch(
        getPostDetails(
          allPosts.find((post) => post.postId === route.params.postId),
        ),
      );
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isFocused]);

  useEffect(() => {
    if (deletePostSuccess) {
      navigation.goBack();
    }

    return () => {
      dispatch(clearPostsDetails({}));
    };
  }, [deletePostSuccess, dispatch, navigation]);

  const handleDeletePost = () => {
    const { postId } = postDetails;
    alertFactory('Delete Post', 'Are you sure you want to delete this post?', [
      {
        text: 'Cancel',
      },
      {
        text: 'Delete',
        onPress: () => dispatch(deletePost(postId, postDetails.imageId)),
      },
    ]);
  };

  const handleEditPost = () => {
    navigation.navigate('UpdatePostDetailsScreen');
    dispatch(isEditingPost(true));
  };

  return (
    <>
      <ScrollView contentContainerStyle={postDetailsStyles.scrollView}>
        <View style={postDetailsStyles.postItemImgContainer}>
          {progressiveImage && (
            <View style={postDetailsStyles.loaderContainer}>
              <Loader />
            </View>
          )}
          <Image
            onLoadStart={() => setProgressiveImageState(true)}
            onLoadEnd={() => setProgressiveImageState(false)}
            style={postDetailsStyles.postItemImg}
            source={{ uri: postDetails.imageUri }}
            resizeMode="cover"
          />
        </View>
        <View style={postDetailsStyles.likesDateCommentsView}>
          <View>
            <Text>{postDetails.likes}</Text>
            <Text>{postDetails.comments}</Text>
          </View>
          <View>
            <Text>{currentDate}</Text>
          </View>
        </View>
        <Text>{postDetails.description}</Text>
      </ScrollView>
      <PostDetailsMenu
        handleDeletePost={handleDeletePost}
        handleEditPost={handleEditPost}
      />
    </>
  );
};

export default PostDetails;
