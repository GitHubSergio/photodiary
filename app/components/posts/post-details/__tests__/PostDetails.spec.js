/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { Alert } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { fireEvent, render, wait } from '@testing-library/react-native';
import reduxMockedStore from '../../../../utils/redux-state';
import PostDetails from '../PostDetails';

jest.mock('@react-navigation/native', () => ({
  useRoute: jest.fn(() => ({
    params: {
      docRef: '4f1dd386-2ebd-485c-95f0-59d9d5d8974e',
      title: 'Copenhagen',
    },
  })),
  useNavigation: () => ({ goBack: jest.fn() }),
  useIsFocused: () => jest.fn(),
}));

jest.mock('react-redux', () => {
  return {
    useSelector: jest.fn(),
    useDispatch: jest.fn(),
  };
});

jest.mock('react-native/Libraries/Alert/Alert.js', () => ({
  alert: jest.fn(),
}));

jest.mock('@react-native-firebase/auth', () => ({}));

jest.mock('@react-native-firebase/firestore', () => ({}));

jest.mock('@react-native-firebase/storage', () => ({}));

describe.skip('<PostDetails />', () => {
  let wrapper;

  beforeEach(() => jest.clearAllMocks());

  it('should render the PostDetails component', async () => {
    useSelector.mockImplementation(() => ({
      postDetails: reduxMockedStore.posts.postDetails,
      success: 200,
    }));
    useDispatch.mockImplementationOnce(() => jest.fn());
    wrapper = render(<PostDetails />);
    await wait(() => wrapper.getByLabelText('submit-button'));
    expect(wrapper.container).toMatchSnapshot();
  });

  it('should dispatch the clearPostSuccess', () => {
    useSelector.mockImplementation(() => ({
      postDetails: {},
      success: 0,
    }));
    const clearPostsSuccess = useDispatch.mockImplementationOnce(() =>
      jest.fn(),
    );
    wrapper = render(<PostDetails />);
    expect(clearPostsSuccess).toHaveBeenCalledTimes(1);
  });

  it('should display the loader when fetching the post details', () => {
    useSelector.mockImplementation(() => ({
      postDetails: {},
      isFetching: true,
    }));
    const clearPostsSuccess = useDispatch.mockImplementationOnce(() =>
      jest.fn(),
    );
    wrapper = render(<PostDetails />);
    expect(clearPostsSuccess).toHaveBeenCalledTimes(1);
    expect(wrapper.container).toMatchSnapshot();
  });

  it('should trigger the Alert modal if the user presses the deletes button', async () => {
    const spy = jest.spyOn(Alert, 'alert');
    useSelector.mockImplementation(() => ({
      postDetails: reduxMockedStore.posts.postDetails,
      isFetching: false,
    }));
    useDispatch.mockImplementationOnce(() => jest.fn());
    wrapper = render(<PostDetails />);
    fireEvent.press(wrapper.getByLabelText('submit-button'));
    await wait(() => Alert.alert.mock.calls[0][2][1].onPress());
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('should delete the post details', async () => {
    useSelector.mockImplementation(() => ({
      postDetails: reduxMockedStore.posts.postDetails,
      isDeleting: true,
    }));
    const deletePost = useDispatch.mockImplementationOnce(() => jest.fn());
    wrapper = render(<PostDetails />);
    expect(deletePost).toHaveBeenCalledTimes(1);
    expect(wrapper.container).toMatchSnapshot();
  });
});
