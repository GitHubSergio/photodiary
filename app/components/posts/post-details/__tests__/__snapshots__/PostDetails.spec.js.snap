// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostDetails /> should delete the post details 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <ScrollView>
    <Image
      source={
        Object {
          "testUri": "../../../app/assets/images/hols1.jpeg",
        }
      }
      style={
        Object {
          "height": 400,
          "width": 400,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    >
      <View>
        <Text>
          10
        </Text>
        <Text>
          10
        </Text>
      </View>
      <View>
        <Text>
          Saturday 22nd Feb 2020
        </Text>
      </View>
    </View>
    <Text>
      Hh
    </Text>
    <View
      testID="loader"
    >
      <ActivityIndicator
        animating={true}
        color="#0000ff"
        hidesWhenStopped={true}
        size="large"
      />
    </View>
  </ScrollView>
</View>
`;

exports[`<PostDetails /> should display the loader when fetching the post details 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    testID="loader"
  >
    <ActivityIndicator
      animating={true}
      color="#0000ff"
      hidesWhenStopped={true}
      size="large"
    />
  </View>
</View>
`;

exports[`<PostDetails /> should render the PostDetails component 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <ScrollView>
    <Image
      source={
        Object {
          "testUri": "../../../app/assets/images/hols1.jpeg",
        }
      }
      style={
        Object {
          "height": 400,
          "width": 400,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    >
      <View>
        <Text>
          10
        </Text>
        <Text>
          10
        </Text>
      </View>
      <View>
        <Text>
          Saturday 22nd Feb 2020
        </Text>
      </View>
    </View>
    <Text>
      Hh
    </Text>
    <TouchableOpacity
      activeOpacity={0.2}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#2D2A4A",
            "borderRadius": 10,
            "height": 40,
            "justifyContent": "center",
          },
          Object {
            "backgroundColor": "#A51931",
            "marginTop": 20,
          },
        ]
      }
    >
      <Text
        accessibilityLabel="submit-button"
        style={
          Object {
            "color": "#F4F5F8",
            "fontSize": 20,
            "fontWeight": "bold",
          }
        }
      >
        DELETE
      </Text>
    </TouchableOpacity>
  </ScrollView>
</View>
`;
